### DISCUSSION TITLE
How do you translate route parameters? (Custom Route Paths)

### DISCUSSION THREAD

**Discussion by MickL** [nuxt-modules/i18n#3028] [General]
Using custom route paths is a nice feature and translating the routes is a kinda must-have for SEO. But how do you guys translate route parameters?

E.g. imagine a clothing shop:

```
File:
/pages/products/[category].vue

EN:
/products/shirts
/products/pants

DE:
/produkte/shirts
/produkte/hosen
```

But how may I handle translated route params? Somehow I need to map it to an enum, e.g. both "pants" and "hosen" should be mapped to enum `Category.Pants`. Of course I could go with the enum as route param instead but then there would be no benefit for SEO: `/products/0` and `/products/1`

**Comment by BobbieGoede** [comment]
I believe what you're asking can be achieved by combining the [`customRoutes` feature](https://i18n.nuxtjs.org/docs/guide/custom-paths) with the [`useSetI18nParams` composable](https://i18n.nuxtjs.org/docs/api#useseti18nparams), which used to set localized [dynamic route parameters](https://i18n.nuxtjs.org/docs/guide/lang-switcher#dynamic-route-parameters).

**Reply by MickL** [reply]
I dont think `useSetI18nParams` is what I am looking for, it seems to be useful for language switching:

* I use `useLocalePath()` to create a menu. Here I need to have products/* be translated, this I could do by translating the enum
* When on a category page the category are already translated oin the current language so I need to reverse-translate them to the Enum before I make the call to the API.

**Comment by MickL** [comment]
@BobbieGoede Some time has passed and v9 is around the corner. I still cant figure out how to solve this, maybe you have an idea?

```vue
<script setup lang="ts">
const route = useRoute();
const category = route.params.id; // <-- This is 'pants' for en, and 'hosen' for de

defineI18nRoute({
  paths: {
    en: '/products/[category]',
    de: '/products/[category]',
  },
});

const { data, pending, error } = await useLazyFetch<ProductsResponse>('/api/products', {
    query: {
      category, // <-- This needs to be 'pants' for all languages
    },
  },
);
</script>
```

If I am not stupid then I need to reverse-translate the category or not? I have a translated string but I need the key. I wonder if no one else has the problem, it should be a very common one.

**Reply by MickL** [reply]
So what I ended up is adding both, the enum AND the translated slug, to my translation. BUT this does not work with `useSwitchLocalePath()` :(

E.g. if I have two categories 'shoes' and 'pants':

de.ts:

```ts
routeSlugs: {
   category: {
      hosen: ProductCategory.Pants,
      [ProductCategory.Pants]: 'hosen' ,
      schuhe: ProductCategory.Shoes,
      [ProductCategory.Shoes]: 'schuhe' ,
   }
}
```

en.ts:

```ts
routeSlugs: {
   category: {
      pants: ProductCategory.Pants,
      [ProductCategory.Pants]: 'pants' , // Error: An object literal cannot have multiple properties with the same name
      shoes: ProductCategory.Shoes,
      [ProductCategory.Shoes]: 'shoes' , // Error: An object literal cannot have multiple properties with the same name
   }
}
```

Then I can get the param this way:

```vue
<script setup lang="ts">
   defineI18nRoute({
      paths: {
         en: '/products/[category]',
         de: '/products/[category]',
      },
   });

   const route = useRoute();
   const categorySlug = route.params.category; // <-- This is 'pants' for en, and 'hosen' for de

   const { t } = useI18n();
   const category: ProductCategory = t(`routeSlugs.category.${categorySlug}`, [], { default: '', missingWarn: false, }) as ProductCategory;

   if (!category) {
      throw createError({
         statusCode: 404,
         fatal: true,
      });
   }

   const { data, pending, error } = await useLazyFetch<ProductsResponse>('/api/products', {
      query: {
         category,
      },
   });
</script>

<template>
   <NuxtLinkLocale
      :to="{ name: 'products-category',
       params: { category: $t(`routeSlugs.category.${ProductCategory.Pants}`) }}"
      >
         $t(`category.${ProductCategory.Pants}`)
      </NuxtLinkLocale>

   <NuxtLinkLocale
      :to="{ name: 'products-category',
       params: { category: $t(`routeSlugs.category.${ProductCategory.Shoes}`) }}"
   >
         $t(`category.${ProductCategory.Shoes}`)
      </NuxtLinkLocale>
</template>
```

This is the simplest approach I came up with. The only other one I can imagine is to write a function that reverse maps the slugs to enum.

@BobbieGoede I think it would make sense for nuxt/i18n to assist with translating slugs. What do you think? Maybe it could provide a slug object similiar to messages that works with fallbacks like messages do:

de-slugs.ts
```ts
{
   name: 'category'
   slugs: {
      {
         slug: 'hosen',
         value: ProductCategory.Pants,
      },
      {
         slug: 'schuhe',
         value: ProductCategory.Shoes,
      }
   }
}
```

en-slugs.ts
```ts
{
   name: 'category'
   slugs: {
      {
         slug: 'pants',
         value: ProductCategory.Pants,
      },
      {
         slug: 'shoes',
         value: ProductCategory.Shoes,
      }
   }
}
```

Then use it somehow like this:

```vue
<script setup lang="ts">
   defineI18nRoute({
      paths: {
         de: '/produkte/[category]',
         en: '/products/[category]',
      },
      slugs: {
         category: 'category', // <-- important for `useSwitchLocalePath()` to work
      }
   });

   const categorySlug = useRoute().params.category;
   const category = useI18n().fromSlug('category', categorySlug);
</script>

<template>
   <NuxtLinkLocale :to="{ name: 'products-category',
       params: { category: $toSlug('category', ProductCategory.Shoes) }}"
   >
      Go to shoes
   </NuxtLinkLocale>
</template>
```

**Reply by MickL** [reply]
So I was thinking more and more about this and I now created an issue that also contains a fully working sandbox example until this feature is implemented into @nuxtjs/i18n: #3409
